### PR TITLE
Tabs component

### PR DESCRIPTION
**Pull request by AleshaOleg**

### CODE REVIEW

**Review by pseudonymcp — CHANGES_REQUESTED** [review]
- Tab selection shouldn't work via `Tab` key, but via arrow keys (right, left). The `Tab` key is preserved for focusing contents within and below the active tab panel, and `Shift`+`Tab` returns the user to the active tab. (-> implementation help: https://inclusive-components.design/tabbed-interfaces/).
- Focus style (focus ring) should only be visible when navigating with the keyboard, not the mouse (-> `focus-visible`).
- The active tab should have `aria-selected` set to `true`.
- The indicator line should have the same width as the text above and not extend it (see design).
- Spacing between tab text and indicator/divider line is too small (see design).
- Tabs in small size show no `cursor: pointer` when hovering the text (only when hovering the small indicator line), but should and behave the same as the other sizes.
- Do we need the `margin` prop? We should use spacing components instead for consitent spacing and keeping our components encapsulated, so I'd remove it.

File: web/src/components/tabs/Tab/index.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      onClick={() => onChangeTab(tab)}
+      onKeyDown={e => onKeyDown(e)}
+      className={classes.tab}
+      {...(isActive && { 'aria-selected': true })}
+      {...(!isActive && { tabIndex: -1 })}
+    >
+      <Text color={isActive ? 'active' : 'primary'} size={getTextSize(size)}>
+        {children}
+      </Text>
+    </a>
```
Comment (tobiaslins) [inline web/src/components/tabs/Tab/index.tsx:75]: Why should the tab be a link?
The generated Uid will change every time so the link change will not do anything.

Just use a button?

File: web/src/components/tabs/Tab/styles.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { createUseStyles } from 'react-jss'
+
+import tokens from '../../../tokens'
+
+export default createUseStyles((theme: typeof tokens) => ({
+  tab: ({ active, align }) => ({
+    background: 'none',
+    zIndex: 1,
```
Comment (tobiaslins) [inline web/src/components/tabs/Tab/styles.ts:8]: do we need that zIndex? it should be avoided if possible

File: web/src/components/tabs/Tabs/index.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import { Tab } from '../Tab'
+
+import useStyles from './styles'
+
+export type TabsProps = {
+  size?: 'large' | 'regular' | 'small'
+  align?: 'left' | 'right' | 'justify'
+  margin?: number
+  width?: number
```
Comment (tobiaslins) [inline web/src/components/tabs/Tabs/index.tsx:12]: do we really want to have a width for a tabs component?
Shouldn't it be always just 100% and the parent limits the width?

If we keep it, it should also allow percentual width (allow string aswell) but I would rather remove it here.

Same for margin. We should think about having the same way to style elements (width, margin) on every component.

Comment (tobiaslins) [reply web/src/components/tabs/Tabs/index.tsx:12]: E.g.: charka-ui has `Box` component that has some basic styling options
https://chakra-ui.com/box

Tab's inherits Box https://chakra-ui.com/tabs

File: web/src/components/tabs/Tabs/index.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  const [active, setActiveTab] = useState(0)
+
+  const onChangeTab = (tab: number) => {
+    if (tab < 0 || tab > children.length - 1) return
+    setActiveTab(tab)
+  }
+
+  return (
+    <div>
```
Comment (tobiaslins) [inline web/src/components/tabs/Tabs/index.tsx:34]: do we really need that extra div wrapped around? please remove!

**Review by tobiaslins — CHANGES_REQUESTED** [review]

File: web/src/components/tabs/Tab/styles.ts (hunk 2 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tab/styles.ts]: > * Tab selection shouldn't work via `Tab` key, but via arrow keys (right, left). The `Tab` key is preserved for focusing contents within and below the active tab panel, and `Shift`+`Tab` returns the user to the active tab. (-> implementation help: https://inclusive-components.design/tabbed-interfaces/).
> * Focus style (focus ring) should only be visible when navigating with the keyboard, not the mouse (-> `focus-visible`).
> * The active tab should have `aria-selected` set to `true`.
> * The indicator line should have the same width as the text above and not extend it (see design).
> * Spacing between tab text and indicator/divider line is too small (see design).
> * Tabs in small size show no `cursor: pointer` when hovering the text (only when hovering the small indicator line), but should and behave the same as the other sizes.
> * Do we need the `margin` prop? We should use spacing components instead for consitent spacing and keeping our components encapsulated, so I'd remove it.

Fixed almost everything from this list, I have some questions

* Can you please design the focus ring? Because there's no default one for `focus-visible`, and I just did something similar to a normal focus ring, but I guess anyway it looks unsuitable.
* No we don't need `margin` prop here, same as `width`. But I added them just to make the component looks pretty in the story. Now I extended the story root block to full size and for me, it looks a bit weird. If it looks ok for you in the story, then we're good to merge it I guess. Otherwise, I can return `width` and `margin` and make it prettier (from my view), as it was before.

**Comment by AleshaOleg** [comment]
> * Tab selection shouldn't work via `Tab` key, but via arrow keys (right, left). The `Tab` key is preserved for focusing contents within and below the active tab panel, and `Shift`+`Tab` returns the user to the active tab. (-> implementation help: https://inclusive-components.design/tabbed-interfaces/).
> * Focus style (focus ring) should only be visible when navigating with the keyboard, not the mouse (-> `focus-visible`).
> * The active tab should have `aria-selected` set to `true`.
> * The indicator line should have the same width as the text above and not extend it (see design).
> * Spacing between tab text and indicator/divider line is too small (see design).
> * Tabs in small size show no `cursor: pointer` when hovering the text (only when hovering the small indicator line), but should and behave the same as the other sizes.
> * Do we need the `margin` prop? We should use spacing components instead for consitent spacing and keeping our components encapsulated, so I'd remove it.

Fixed almost everything from this list, I have some questions

* Can you please design the focus ring? Because there's no default one for `focus-visible`, and I just did something similar to a normal focus ring, but I guess anyway it looks unsuitable.
* No we don't need `margin` prop here, same as `width`. But I added them just to make the component looks pretty in the story. Now I extended the story root block to full size and for me, it looks a bit weird. If it looks ok for you in the story, then we're good to merge it I guess. Otherwise, I can return `width` and `margin` and make it prettier (from my view), as it was before.

File: web/src/components/tabs/Tab/index.tsx (hunk 1 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tab/index.tsx]: Yes, just did the link firstly as it was in the tutorial. forgot to change back to the button. Though not everything will work button and used a link for testing. Thanks!

File: web/src/components/tabs/Tabs/index.tsx (hunk 3 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tabs/index.tsx:12]: I described why I added `margin` and `width` here - https://github.com/ubergrape/grape-ds/pull/33#issuecomment-720353906

I removed them for now because in the reality of usage they really don't have any sense.

File: web/src/components/tabs/Tabs/index.tsx (hunk 4 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tabs/index.tsx]: Removed, thanks. Added this to push blocks in stories to a new line :D

File: web/src/components/tabs/Tab/styles.ts (hunk 2 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tab/styles.ts]: Unfortunately, I don't see any other solution here, how I can overlay a dark grey indicator line for the tab on hover

**Comment by pseudonymcp** [comment]
Looks mostly fine, I just noticed the following:
- Width and alignment of divider and indicator line are off (see image)
- Is space at the top missing? The tab group "small" should have 24px height, the tab groups "regular" and "large" should have 32px height

Tabs-Code-Design

- I also added `focus-ring` with adjusted spacing to component design:   https://www.figma.com/file/Mc03kFcaFlcMOA7afjaZmF/Grape-DS-Components-Web?node-id=705%3A2559

No margin and width is fine, we don't need to tweak them to look good in Storybook, this is just for checking the integrity of the components.

**Comment by pseudonymcp** [comment]
@AleshaOleg  Looks mostly fine for me now, just two more things:
- When set to "justified", the tab expands in width with the text centered. The spacing between the tabs stays the same for every alignment. So the indicator should stretch the whole tab width when set to "justified", not just be placed below the text. 

image

- The is an accessibility issue (it's also reported in Storybook in the Accessibility panel): The tab is a `button` with `aria-selected` set to `true`. This is not allowed as for buttons aria-selected is not allowed. We need to define its role as tab.
More info on that and hints to solve:
- https://www.stefanjudis.com/blog/aria-selected-and-when-to-use-it/
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role

More on role hierarchy (what's allowed for which role): https://www.w3.org/WAI/PF/aria/rdf_model.svg

**Review by pseudonymcp — CHANGES_REQUESTED** [review]
@AleshaOleg  Looks mostly fine for me now, just two more things:
- When set to "justified", the tab expands in width with the text centered. The spacing between the tabs stays the same for every alignment. So the indicator should stretch the whole tab width when set to "justified", not just be placed below the text. 

image

- The is an accessibility issue (it's also reported in Storybook in the Accessibility panel): The tab is a `button` with `aria-selected` set to `true`. This is not allowed as for buttons aria-selected is not allowed. We need to define its role as tab.
More info on that and hints to solve:
- https://www.stefanjudis.com/blog/aria-selected-and-when-to-use-it/
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role

More on role hierarchy (what's allowed for which role): https://www.w3.org/WAI/PF/aria/rdf_model.svg

**Review by pseudonymcp — APPROVED** [review]
Looks good for me now 👍

File: web/src/components/tabs/Tab/index.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    case 'large':
+      return 'body-large'
+    case 'regular':
+      return 'body-base'
+    case 'small':
+      return 'body-small'
+    default:
+      return 'body-base'
+  }
+}
```
Comment (tobiaslins) [inline web/src/components/tabs/Tab/index.tsx:19]: Ok for now -> we will change `Text` size prop to 'small'|'regular'|'large' aswell!

Will do this in a separate PR after things are merged

File: web/src/components/tabs/Tabs/styles.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    position: 'relative',
+    '&::after': {
+      content: '""',
+      position: 'absolute',
+      bottom: 0,
+      left: 8,
+      width: 'calc(100% - 16px)',
+      height: theme.borderWidthTab,
+      backgroundColor: theme.colorBorderTabDivider,
+    },
```
Comment (tobiaslins) [inline web/src/components/tabs/Tabs/styles.ts:28]: This is just because of the bottom line right?

This can be replaced by adding a box shadow to the `wrapper`

`box-shadow: inset 0px -2px 0px rgb(241, 241, 241);`

This will also resolve the zIndex "issue"

The first child should not have padding on the left to keep styles aligned

File: web/src/components/tabs/Tab/styles.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import tokens from '../../../tokens'
+
+export default createUseStyles((theme: typeof tokens) => ({
+  button: {
+    background: 'none',
+    zIndex: 1,
+    display: 'flex',
+    alignItems: 'stretch',
+    position: 'relative',
+    padding: [0, '8px', 0, '8px'],
```
Comment (tobiaslins) [inline web/src/components/tabs/Tab/styles.ts:12]: Use `size1X` instead of `8px`

File: web/src/components/tabs/Tab/styles.ts (hunk 2 above)
Comment (tobiaslins) [reply web/src/components/tabs/Tab/styles.ts:8]: See other comment of `:after` element

**Review by tobiaslins — CHANGES_REQUESTED** [review]
We need to discuss a way on how to actually **use** tabs.

Currently, they are just rendering the tabs. How are they used to change content?
Should there be a way to a "Controlled" tab component? @pseudonymcp 

e.g.: 

```jsx
<Tabs>
  <Tab name="Tab name 1">
    <div>My tab body content</div>
  </Tab>
  <Tab name="Tab name 2">
    <div>My tab body content 2</div>
  </Tab>
</Tabs>
```

File: web/src/components/tabs/Tabs/styles.ts (hunk 6 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tabs/styles.ts:28]: If the first and last child will not have padding on the left and right the outline on focus will look weird. Is it ok @pseudonymcp?

File: web/src/components/tabs/Tab/styles.ts (hunk 7 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tab/styles.ts:12]: Thanks!

File: web/src/components/tabs/Tabs/styles.ts (hunk 6 above)
Comment (AleshaOleg) [reply web/src/components/tabs/Tabs/styles.ts:28]: Screenshot 2020-11-04 at 11 54 25

will look, like this

**Review by pseudonymcp — APPROVED** [review]

**Comment by pseudonymcp** [comment]
Looks good for me now 👍

**Comment by AleshaOleg** [comment]
> We need to discuss a way on how to actually **use** tabs.
> 
> Currently, they are just rendering the tabs. How are they used to change content?
> Should there be a way to a "Controlled" tab component? @pseudonymcp
> 
> e.g.:
> 
> ```js-jsx
> <Tabs>
>   <Tab name="Tab name 1">
>     <div>My tab body content</div>
>   </Tab>
>   <Tab name="Tab name 2">
>     <div>My tab body content 2</div>
>   </Tab>
> </Tabs>
> ```

Just added this functionality. Can you please review this PR again, so we can finally merge it? Thanks!

File: web/src/stories/tabs/base.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    <Tab>Tab name 1</Tab>
+    <Tab>Tab name 2 </Tab>
+    <Tab>Tab name 3</Tab>
+    <TabPanels>
+      <div>Tab 1 content</div>
+      <div>Tab 2 content</div>
+      <div>Tab 3 content</div>
+    </TabPanels>
+  </Tabs>
+)
```
Comment (tobiaslins) [inline web/src/stories/tabs/base.stories.tsx:26]: Can we move to the proposed API?
This feels unnatural to use! 
```
<Tabs>
  <Tab name="Tab name 1">
    <div>My tab body content</div>
  </Tab>
  <Tab name="Tab name 2">
    <div>My tab body content 2</div>
  </Tab>
</Tabs>
```

This way it's clear what tab contains what content. It's how other design systems is exposing it! Also, one component less to import :)

**Review by tobiaslins — CHANGES_REQUESTED** [review]
One last time :) I can also do it if you want

File: web/src/stories/tabs/base.stories.tsx (hunk 8 above)
Comment (AleshaOleg) [reply web/src/stories/tabs/base.stories.tsx]: Agree, just pushed changes
